reducers/taskReducer.js

```diff
@@ -0,0 +1,73 @@
+import {
+  ADD_TASK,
+  REMOVE_TASK,
+  TOGGLE_COMPLETE_TASK,
+  MOVE_TASK
+} from "../actions/taskActions";
+
+export default (state = {}, action) => {
+  switch (action.type) {
+    case ADD_TASK: {
+      const newTasks = { ...state };
+      const newTaskListForDate = [];
+      if (newTasks[action.selectedDate]) {
+        newTaskListForDate.push(...newTasks[action.selectedDate]);
+      }
+      newTaskListForDate.unshift({
+        ...action.task,
+        completed: false,
+        // TODO: figure out a better pattern for ID generation
+        id: String(Date.now())
+      });
+      newTasks[action.selectedDate] = newTaskListForDate;
+      return {
+        ...state,
+        ...newTasks
+      };
+    }
+    case REMOVE_TASK: {
+      const { selectedDate, id } = action;
+      const newTasks = { ...state };
+      const prevIndex = newTasks[selectedDate].findIndex(
+        item => item.id === id
+      );
+      newTasks[selectedDate].splice(prevIndex, 1);
+      return {
+        ...state,
+        ...newTasks
+      };
+    }
+    case TOGGLE_COMPLETE_TASK: {
+      const newTasks = { ...state };
+      newTasks[action.selectedDate].find(
+        t => t.id === action.task.id
+      ).completed = action.task.completed;
+      return {
+        ...state,
+        ...newTasks
+      };
+    }
```

Comment:
This approach should work fine for the list view, where we're showing the list of tasks per day.

However, in the full production app, this would need to be altered. If you're viewing the 'details view' (the view that appears when you select a task from the list), you can perform actions from there as well (e.g. move, complete). From that details view, you aren't guaranteed a selectedDate - the same task might appear on multiple days.

Again, what you've done here is totally fine for the scope of the project, just letting you know about extra complexity that would need to be addressed were this to be implemented for reals

Reply:
Thats a great point... I think basically all of these reducers could be refactored to look up task by id alone instead of looking under the `selectedDate` key which would support the functionality you are mentioning. 

Something to consider as a future enhancement